### PR TITLE
REF: use fused types in algos_rank_helper

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 1]
```diff
@@ -8,110 +8,104 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
...
     # double sort first by mask and then by values to ensure nan values are
     # either at the beginning or the end. mask/(~mask) controls padding at
     # tail or the head
-    {{if dtype != 'uint64'}}
-    if ascending ^ (na_option == 'top'):
-        nan_value = {{pos_nan_value}}
-        order = (values, mask)
+    if rank_t is not uint64_t:
+        if ascending ^ (na_option == 'top'):
+            if rank_t is object:
```
Comment (WillAyd) [inline pandas/_libs/algos_rank_helper.pxi.in]: Not necessarily in scope for this but would be nice if there was a better way to do this via fused types (a dict or inline func?)

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 2]
```diff
@@ -8,110 +8,104 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
...
+    if rank_t is object:
         _as = np.lexsort(keys=order)
-        if not ascending:
-            tiebreak = TIEBREAK_FIRST_DESCENDING
     else:
-        _as = np.lexsort(keys=order)
-    {{endif}}
+        if tiebreak == TIEBREAK_FIRST:
+            # need to use a stable sort here
+            _as = np.lexsort(keys=order)
```
Comment (WillAyd) [inline pandas/_libs/algos_rank_helper.pxi.in]: Maybe reading wrong but this is duplicated in each conditional branch? Could clean in follow up if so

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 3]
```diff
@@ -122,38 +116,30 @@ def rank_1d_{{dtype}}({{ctype}}[:] in_arr, ties_method='average',
...
-    {{endif}}
-        # TODO: why does the 2d version not have a nogil block?
+    if rank_t is object:
+        # TODO: de-duplicate once cython supports conditional nogil
         for i in range(n):
             sum_ranks += i + 1
             dups += 1
 
-            {{if dtype == 'object'}}
-            val = util.get_value_at(sorted_data, i)
```
Comment (WillAyd) [inline pandas/_libs/algos_rank_helper.pxi.in]: Is this line no longer required?

Comment (jbrockmendel) [reply pandas/_libs/algos_rank_helper.pxi.in]: AFAICT it hasnt been needed for a while.  I recently removed most usages of util.get_value_at as unnecessary when we know `i` is an integer.

File: pandas/_libs/algos_rank_helper.pxi.in (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_rank_helper.pxi.in]: agreed

File: pandas/_libs/algos_rank_helper.pxi.in (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_rank_helper.pxi.in]: Yah I've been giving that some thought.  I think we can do that with something like:

```
cdef inline rank_t get_thing(rank_t val):  # <-- totally ignore `val`, just care about rank_t
    if rank_t is object: ...

def actual_func(rank_t ...):
    cdef:
          rank_t val, thing
    
    thing = get_thing(val)
```

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 4]
```diff
@@ -8,110 +8,104 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
...
+    if rank_t is not uint64_t:
+        if ascending ^ (na_option == 'top'):
+            if rank_t is object:
+                nan_value = Infinity()
+            elif rank_t is float64_t:
+                nan_value = np.inf
+            elif rank_t is int64_t:
+                nan_value = np.iinfo(np.int64).max
+
+            order = (values, mask)
```
Comment (jreback) [inline pandas/_libs/algos_rank_helper.pxi.in]: can you add some comments here

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 5]
```diff
@@ -231,103 +269,106 @@ def rank_2d_{{dtype}}({{ctype}}[:, :] in_arr, axis=0, ties_method='average',
...
-    mask = np.isnan(values)
-    {{elif dtype == 'int64'}}
-    mask = values == NPY_NAT
-    {{endif}}
+    if rank_t is not uint64_t:
+        if ascending ^ (na_option == 'top'):
+            if rank_t is object:
+                nan_value = Infinity()
+            elif rank_t is float64_t:
+                nan_value = np.inf
```
Comment (jreback) [inline pandas/_libs/algos_rank_helper.pxi.in]: comments here as well

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. just some small requests, i suppose you have some followups to de-dupliacate

File: pandas/_libs/algos_rank_helper.pxi.in (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_rank_helper.pxi.in]: can you be more specific?  there is a pretty nice comment up on L69

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**